### PR TITLE
SqueezeNet model with functional API 

### PR DESCRIPTION
**Pull request by rcmalli**
- Squeezenet has very small footprint ( 5.1 MB model file ).
- It has AlexNet accuracy.
- Feedforward prediction time is near 5 FPS with CPU (2.4 GHz Intel Core i5).
- This project contains both Tensorflow and Theano weights ported from Caffe weights.
- Weights are placed in models folder at [keras-squeezenet](https://github.com/rcmalli/keras-squeezenet) repository.
- Code is rewritten and follows the rules of other model codes in keras/applications.

### TO-DO

- Links for the model files should be fixed according to release version of this library (look at squeezenet.py).

### Reference:
- [SqueezeNet: AlexNet-level accuracy with 50x fewer parameters and <0.5MB model size](https://arxiv.org/abs/1602.07360)

### Keras Project Reference:

- [keras-squeezenet](https://github.com/rcmalli/keras-squeezenet)

### Original Project Reference:

- [Original Squeezenet](https://github.com/DeepScale/SqueezeNet)

### CODE REVIEW

File: squeezenet.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+# Keras Project Reference:
+
+- [keras-squeezenet](https://github.com/rcmalli/keras-squeezenet)
+
+# Original Project Reference:
+
+- [Original Squeezenet](https://github.com/DeepScale/SqueezeNet)
+
+'''
```
Comment (fchollet) [inline squeezenet.py]: The docstring should come at the top of the file.

File: squeezenet.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+TH_WEIGHTS_PATH = 'PATH/squeezenet_weights_th_dim_ordering_th_kernels.h5'
+TF_WEIGHTS_PATH = 'PATH/squeezenet_weights_tf_dim_ordering_tf_kernels.h5'
```
Comment (fchollet) [inline squeezenet.py]: Where can the weights be downloaded?

File: squeezenet.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    right = Convolution2D(expand, 3, 3, border_mode='same', name=s_id + exp3x3)(x)
+    right = Activation('relu', name=s_id + relu + exp3x3)(right)
+
+    x = merge([left, right], mode='concat', concat_axis=c_axis, name=s_id + 'concat')
+    return x
+
+
+def SqueezeNet(nb_classes, finetune=False, input_tensor=None, weights=None):
+    '''Instantiate the SqueezeNet architecture,
+        optionally loading weights pre-trained
```
Comment (fchollet) [inline squeezenet.py]: Indentation issue?

File: squeezenet.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    left = Activation('relu', name=s_id + relu + exp1x1)(left)
+
+    right = Convolution2D(expand, 3, 3, border_mode='same', name=s_id + exp3x3)(x)
+    right = Activation('relu', name=s_id + relu + exp3x3)(right)
+
+    x = merge([left, right], mode='concat', concat_axis=c_axis, name=s_id + 'concat')
+    return x
+
+
+def SqueezeNet(nb_classes, finetune=False, input_tensor=None, weights=None):
```
Comment (fchollet) [inline squeezenet.py]: Shouldn't `weights` default to `"imagenet"`?

Comment (fchollet) [inline squeezenet.py]: In general: please match the same constructor arguments as every other image classification model in the repo / in `keras.applications`: `(include_top=True, weights='imagenet', input_tensor=None)`

**Comment by rcmalli** [comment]
- Docstring placement and indentation are fixed.
- Download links for weights are added.
- Function arguments are changed.

**Comment by fchollet** [comment]
Sorry for the delay. I do intend on merging this, I just have limited bandwidth.

**Comment by soupault** [comment]
@rcmalli do you think if it would be a good idea to add also head-less versions of the architecture (see, e.g. ResNet50) so to ease the potential usage in transfer learning?
